[Backtracing][Linux] Add support for Linux in the _Backtracing module.

al45tair opened the pull request:
Rearrange things to avoid using SwiftShims, because that's grotty and means we have to carry all kinds of definitions from the C library. Instead, turn off implicit modules for the `_Backtracing` build and add some explicit modules that we can use as we please.

Also add the code to handle ELF and DWARF for Linux, including symbol lookup, demangling, line number/source file extraction, Linux `mcontext_t` support et al.

Tests for all of this will be in a separate PR that also enables everything.

rdar://110261712

al45tair committed:
[Backtracing][IRGen] Add a semantic attribute to force frame pointer.

The Swift backtracer's frame pointer unwinder cannot work on Linux
without this change, because the compiler omits the frame pointer from
the function in libSwift_Backtracing that actually captures the stack.

rdar://110260855

al45tair committed:
[Backtracing][Linux] Add Linux crash handler to the runtime.

This also adds a function to demangle a symbol, and a way for the
backtracing code to report warning messages to the same place as
the main runtime.

I'd like to rename the _swift_isThunkFunction() SPI also, but we
can't do that until we've made the changes to the _Backtracing
library, so we'll do that there instead.

rdar://110261430

al45tair committed:
[Backtracing][Linux] Reorganise modules for _Backtracing.

Using `SwiftShims` is undesirable - it creates all kinds of build issues,
and means shipping the `_SwiftBacktracing.h` header in the SDK, which is
not necessary.

While we're doing this, add the necessary definitions for reading ELF
and DWARF information.

rdar://110261712

al45tair committed:
[Backtracing][Linux] Add Linux support to the _Backtracing module.

Use the new module structure rather the old SwiftShims header.  This
is much cleaner and lets us include operating system headers to get
the relevant definitions where possible.

Add code to support ELF and DWARF, including decompression using
zlib, zstd and liblzma if those turn out to be required and available.

rdar://110261712

al45tair commented:
This depends on #66333 and #66334.

al45tair committed:
[Backtracing][Linux] Add missing CrashInfo.h header.

This was added to a later PR, but not to this one, though we need
it here.

rdar://110261430

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Backtracing][Linux] Disable backtrace-on-crash until enabling PR.

This should have been disabled until #66338.

rdar://110261430

al45tair committed:
[Backtracing][Linux] Temporarily add a default addressWidth.

This is for compatibility, so that I can split up the PRs.
We'll remove it in the next PR.

rdar://110261712

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test Linux platform

al45tair committed:
[Backtracing][Linux][Tests] Fix verify_all_overlays test.

The `_Backtracing` module has a number of private implementation only
imports that aren't used outside of the module and that don't require
any additional libraries (hence they aren't relevant to the outside
world).  `verify_all_overlays.py` needs to know about these when it
does its test, because it loadas the module as the main module, which
results in implementation only imports being required instead of
ignored.

rdar://110261712

al45tair commented:
When reviewing this, you can start at 7e3df6e; the changes prior to that are in #66333 and #66334 and will be reviewed there instead.

al45tair commented:
If you're wondering where the tests are, I left them until the last PR (#66338) so everything was enabled and they'd actually work.

mikeash commented on stdlib/public/Backtracing/ArrayImageSource.swift:
```diff
@@ -0,0 +1,50 @@
+        throw ArrayImageSourceError.outOfBoundsRead(addr, requested)
+      }
+
+      memcpy(buffer.baseAddress!, $0.baseAddress! + Int(addr), Int(requested))
```
You could avoid the libc dependency by doing something like `buffer.update(withContentsOf: $0[addr ..< addr + requested]`. Not that it matters much.

al45tair replied:
I did try something like that, but got a bit tangled up because of types (`U` isn't necessarily `UInt8` here) and decided this was the clearest way to write what I meant. It'd be nice to revisit it at some point, but this works for now.

mikeash commented on stdlib/public/Backtracing/MemoryReader.swift:
```diff
@@ -60,55 +76,174 @@ extension MemoryReader {
+                       bytes.baseAddress!.advanced(by: done),
+                       Int(reply.len))
+
+        if ret < 0 || ret != reply.len {
```
Another potential partial read.

al45tair replied:
Yeah, while I haven't seen it happen, I think you're right that I should handle that.

mikeash reviewed (approved):
LGTM. I'm trusting you and your tests for large chunks of the DWARF/ELF/compression stuff.

al45tair committed:
[Backtracing][Linux] Tidy a few things up after review.

Mike and Max made various helpful suggestions, so I've added and updated
various comments and amended the code to cope with partial reads and
writes.

rdar://110261430

al45tair committed:
[Backtracing][Linux] Move doc comment to header.

Moved the comment for `_swift_backtrace_demangle` into the header file
instead of it being in the implementation.

rdar://110261430

al45tair committed:
[Backtracing][Linux] Work correctly in the presence of partial reads.

There's a chance that pipes might perform a partial read; we should
handle that case.

rdar://110261712

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

etcwilde commented on include/swift/Runtime/Backtrace.h:
```diff
@@ -123,8 +134,42 @@ struct BacktraceSettings {
+///
+/// @returns a pointer to the output if demangling was successful.
+SWIFT_RUNTIME_STDLIB_SPI
+char *_swift_backtrace_demangle(const char *mangledName,
```
Do we want to indicate whether it was a swift symbol or C++ symbol that was demangled? The status is a little weird otherwise? And do we want to document the values of status?

e.g. (no idea what Windows does here though so... ¯\\\_(ツ)\_/¯)

```
*status is set to one of the values:
 0: The demangling operation succeeded.
-1: A memory allocation failiure occurred.
-2: mangled_name is not a valid name under the C++ ABI mangling rules.
-3: One of the arguments is invalid.
```

etcwilde replied:
(I guess the `-2` value should read that the `mangled_name` is not valid under the C++ ABI or Swift mangling rules)

al45tair replied:
Windows is totally different, helpfully.

etcwilde replied:
😒 Windows... Are we going to be internally consistent and always returning the same set of status numbers for the various kinds of failures, or are we going to be externally consistent and return... a number?

al45tair replied:
I'm not too bothered about this, as it's SPI that's only used by the backtracing implementation (which is built in lock-step with the runtime and right now actually ignores `status` anyway — it's really only there for debugging; if we can't demangle something, we just don't), and I don't even mind the signature changing between Windows and Linux. macOS isn't using this function.

etcwilde commented on stdlib/public/runtime/Backtrace.cpp:
```diff
@@ -789,20 +843,100 @@ namespace backtrace {
+
+  // You must provide buffer size if you're providing your own output buffer
+  if (outputBuffer && !outputBufferSize) {
+    return nullptr;
```
Would it make sense to set `*status = -3` here as well?

al45tair replied:
I'm not really proposing `status` as any kind of useful interface. I just wanted it available for debugging and never removed it.

etcwilde commented on stdlib/public/runtime/Backtrace.cpp:
```diff
@@ -789,20 +843,100 @@ namespace backtrace {
+  } else if (name.startswith("_Z")) {
+    // Try C++
+    size_t resultLen;
+    char *result = abi::__cxa_demangle(mangledName, nullptr, &resultLen, status);
```
Doesn't `abi::__cxa_demangle` pretty much define the same behavior? Do we need the wrapper checks remapping the results? (It doesn't check the mangledNameLength, which I would appreciate)
But something like this
`return abi::__cxa_demangle(mangledName, outputBuffer, outputBufferSize, status);` looks like it should do the same thing, but without the additional memcpy. I suppose this doesn't work if the `outputBuffer` isn't allocated with `malloc` though. :thinking:. (If that's the reason why we're stashing the pointer and copying the result manually, that would be good to comment so someone doesn't come along and try reducing it.)

Which reminds me, we should probably document that the result pointer needs to be free'd with `free` in either case.

al45tair replied:
Yeah, I did that so that you didn't have to use `malloc()` to allocate the buffer. One of the nasty things about the `abi::__cxa_demangle` function is that it requires that the pointer you pass in is `malloc()`ed and might even `realloc()` it for you. I can see how that came about, but I can't say I'm keen on it as an interface.

etcwilde replied:
yeah, okay. Using `realloc` internally.... That's an interesting decision. Not a huge fan, but oh well.

etcwilde replied:
Anyway comment if you could add a comment to the call location, I'm fine with this how it is then.

etcwilde reviewed (commented):
Okay, some comments. Maybe could use a little more documenting. The code mostly looks fine though.

etcwilde commented on stdlib/public/runtime/Backtrace.cpp:
```diff
@@ -789,20 +843,100 @@ namespace backtrace {
+    }
+
+    if (outputBuffer == nullptr) {
+      outputBuffer = (char *)::malloc(result.length() + 1);
```
I think this needs a pointer check to make sure it's not a `nullptr`.

al45tair replied:
I'm normally of the "don't bother checking if `malloc()` returned `NULL`" school of thought, especially on Linux where the allocator overcommits (at least, unless you force that off) and where _actually_ running out of memory might cause the OOM killer to arbitrarily victimise something anyway.

I guess we _could_ check and return _nullptr_, but we're probably in such a mess already if that happens that there's little hope of any good outcome.

etcwilde replied:
Isn't it set `*status` to `-1` in that case?

al45tair replied:
Ah, I see what you're talking about. Maybe I should just remove `status`, since we don't use it anyway.

al45tair committed:
[Backtracing][Linux] Remove status argument.

The `status` argument to the `_swift_backtrace_demangle()` function
isn't especially useful, won't match behaviour on Windows, and we
actually don't use it in the Swift code that calls this SPI.

Remove it.

rdar://110261712

al45tair committed:
[Backtracing][Linux] Add a comment about __cxa_demangle.

`__cxa_demangle()` is a rather unusual API; one of its "features" is that
the pointer you pass in must either be `nullptr`, in which case it
will call `malloc()` itself, _or_ it has to be a pointer to a block
of memory allocated with `malloc()`, because `__cxa_demangle()` may
`realloc()` it for you.

This seems to me to be something of a non-obvious footgun, so we never
pass the caller's pointer through to `__cxa_demangle()`, which lets them
decide how they want to allocate space.

rdar://110261712

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Backtracing][Linux] Also remove the status argument from Runtime module.

There's a separate declaration here because we can't include the `Backtrace.h`
header from inside `modules/Runtime/Runtime.h`.

rdar://110261712
